### DISCUSSION TITLE
Difference between `Option` and `Either`?

### DISCUSSION THREAD

**Discussion by lonix1** [louthy/language-ext#998] [Q&A] [answered]
How do `Option` and `Either` differ?

Similar libraries have a `Maybe` concept as well - is that the same as the former or latter?

**Comment by StefanBertels** [comment, marked as answer]
`Option` is "Maybe" -- which means `None` or `Some(value)` ([sometimes ](https://en.wikibooks.org/wiki/Haskell/Understanding_monads/Maybe) called Nothing/Just).
* [Introduction to Option in LanguageExt](https://github.com/louthy/language-ext#option)
* [Documentation for Option monad](https://louthy.github.io/language-ext/LanguageExt.Core/Monads/Alternative%20Value%20Monads/Option/OptionUnsafe/index.html)

`Either` means "one of two values" 
* [Documentation for Right/Left choice monad](https://louthy.github.io/language-ext/LanguageExt.Core/Monads/Alternative%20Value%20Monads/Either/Either/index.html)

**Reply by lonix1** [reply]
Your one-liner definitions made everything "click" - the docs are great  but a tl;dr summary like yours would be helpful for each section. Thank you!

**Reply by StefanBertels** [reply]
I recommend reading some introduction. This is a good book:
https://www.manning.com/books/functional-programming-in-c-sharp-second-edition

https://github.com/louthy/language-ext and https://github.com/louthy/language-ext/wiki IMO are very helpful, too.

**Comment by CK-LinoPro** [comment]
One thing to add: `Either<Unit, T>` is structurally the same as `Option<T>`.

You can think of types in terms of their cardinality for comparison, i.e. their number of possible values.
Two types with the same cardinality have the same range of values and you can define a lossless conversion (isomorphism) between the two.
If two types have different cardinality, then one type holds more information than the other.

``` cs
Card(Option<T>) == Card(T) + 1
Card(Either<L, R>) == Card(L) + Card(R)
Card(Unit) == 1
Card(Either<Unit, T>) == Card(T) + 1
```

**Reply by lonix1** [reply]
Never thought of types that way before, very interesting, thank you!

**Reply by CK-LinoPro** [reply]
Just to give you an impression of how far this rabbit hole goes: this view of cardinality allows you to perform calculus on types, which is how zippers are derived.
Zippers are data structures with positional information.

E.g. a list is classically defined as `Seq<T> = Empty | Cons(T head, Seq<T> tail)` (pseudo-syntax).
It has a cardinality of `Card(Seq<T>) = Card(T) + Card(T) * Card(Seq(T)) = 1/(1-Card(T))`.
The derivative is `d/d(Card(T)) Card(Seq<T>) = Card(Seq<T>) * Card(Seq<T>)`,
which leads to the "hole type" `SeqHole<T> = (Seq<T> previous, Seq<T> next)` lacking the currently selected element,
which again leads to the zipper type `SeqZipper<T> = (Seq<T> previous, Option<T> current, Seq<T> next)`, consisting of the currently selected value (if any), a list of previous values and a list of following values.

This isn't anything you'll need very soon though.

https://en.wikibooks.org/wiki/Haskell/Zippers
